### PR TITLE
IPython kernel plugin

### PR DESCRIPTION
**Pull request by dpinte**
@pberkes This is the code we talked about.

You can test the integration using `python examples/plugings/tasks/ipython_kernel/example.py`

If time permits, I will add some feature to streamline contributions from the app to the kernel namespace.

### CODE REVIEW

**Comment by pberkes** [comment]
@dpinte Thanks for this, this week I'm teaching, I'll review next week

**Comment by pberkes** [comment]
Are there any restrictions in terms of IPython version number? It works on 2.4.1 and 3.1.0.

**Comment by pberkes** [comment]
In 3.1.0, `connect_qt_console` has moved to `IPython.kernel.connect` from `IPyhton.lib.kernel`

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from IPython.kernel.zmq.kernelapp import IPKernelApp
+
+
+def mpl_kernel(gui):
+    """ Launch and return an IPython kernel with matplotlib support for the
+    desired gui.
+
+    """
+    kernel = IPKernelApp.instance()
+    kernel.initialize(['python', '--pylab=%s' % gui])
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: How about using `--maplotlib=%s`?

**Comment by pberkes** [comment]
When the example application closes, the consoles are left open. The kernel has been killed, so they are basically frozen. 

It would be nice to add the code to close them cleanly on exit, as reference for real apps.

File: envisage/plugins/ipython_kernel/actions.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+from traits.api import Instance
+
+from .internal_ipkernel import InternalIPKernel
+
+
+class StartQtConsoleAction(TaskAction):
+
+    id = 'ipython_qtconsole'
+
+    name = 'Console'
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/actions.py]: "IPython Console"?

File: envisage/plugins/ipython_kernel/ipython_kernel_plugin.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    #### Trait initializers ###################################################
+
+    def _contributed_task_extensions_default(self):
+
+        from .actions import StartQtConsoleAction
+
+        def menu_factory():
+            return SGroup(
+                StartQtConsoleAction(kernel=self.kernel),
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/ipython_kernel_plugin.py]: This relies on `self.kernel` having been created already. We should grab it from the service, or else have a `_get_kernel` method that creates a kernel if necessary.

File: examples/plugins/tasks/ipython_kernel/example.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            kernel = self.get_service(IPYTHON_KERNEL_PROTOCOL)
+
+            # Create windows from the default or saved application layout.
+            self._create_windows()
+
+            app.connect(app, QtCore.SIGNAL("lastWindowClosed()"),
+                        app, QtCore.SLOT("quit()"))
+
+            app.aboutToQuit.connect(kernel.cleanup_consoles)
```
Comment (pberkes) [inline examples/plugins/tasks/ipython_kernel/example.py]: This does not happen: I suspect the console windows themselves prevent the Qt application from quitting. We could listen to the `application_exiting` event instead.

BTW shouldn't we call `kernel.shutdown` instead?

**Comment by dpinte** [comment]
@pberkes thanks for the review. 

For reference, this is a useful link: https://github.com/ipython/ipython/issues/7614

File: examples/plugins/tasks/ipython_kernel/i_python_editor.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in enthought/LICENSE.txt and may be redistributed only
+# under the conditions described in the aforementioned license.  The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+# Thanks for using Enthought open source!
+#
+# Author: Enthought, Inc.
+# Description: <Enthought pyface package component>
+#------------------------------------------------------------------------------
```
Comment (pberkes) [inline examples/plugins/tasks/ipython_kernel/i_python_editor.py]: Outdated headers

File: examples/plugins/tasks/ipython_kernel/python_editor.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+# Copyright (c) 2007, Riverbank Computing Limited
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD license.
+# However, when used with the GPL version of PyQt the additional terms described in the PyQt GPL exception also apply
+
+#
+# Author: Riverbank Computing Limited
+# Description: <Enthought pyface package component>
+#------------------------------------------------------------------------------
```
Comment (pberkes) [inline examples/plugins/tasks/ipython_kernel/python_editor.py]: Outdated header

File: envisage/plugins/ipython_kernel/ipython_kernel_plugin.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    # The plugin unique identifier.
+    id = 'envisage.plugins.ipython_kernel'
+
+    # The plugin name (suitable for displaying to the user).
+    name = 'IPython embedded kernel plugin'
+
+    def stop(self):
+        logger.info('Shutting down the embedded ipython kernel')
+        self.kernel.shutdown()
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/ipython_kernel_plugin.py]: Mmmh this should have closed the console windows, right?

File: envisage/plugins/ipython_kernel/ipython_kernel_plugin.py (hunk 3 above)
Comment (pberkes) [inline envisage/plugins/ipython_kernel/ipython_kernel_plugin.py]: Actually, the easiest is to define a `kernel` trait with a default initializer.

File: examples/plugins/tasks/ipython_kernel/example.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+            kernel.init_ipkernel('qt4')
+
+            kernel.namespace['app'] = self
```
Comment (pberkes) [inline examples/plugins/tasks/ipython_kernel/example.py]: The `namespace` attribute should be set by an extension point.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        # Start IPython kernel with GUI event loop and mpl support
+        self.ipkernel = mpl_kernel(gui_backend)
+        # To create and track active qt consoles
+        self.consoles = []
+
+        # This application will also act on the shell user namespace
+        self.namespace = self.ipkernel.shell.user_ns
+
+    def new_qt_console(self, evt=None):
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: `evt` is not used

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        """ Start a new qtconsole connected to our kernel. """
+        return connect_qtconsole(
+            self.ipkernel.connection_file, profile=self.ipkernel.profile
+        )
+
+    def cleanup_consoles(self, evt=None):
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: `evt` is not used

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        self.consoles = []
+
+        # This application will also act on the shell user namespace
+        self.namespace = self.ipkernel.shell.user_ns
+
+    def new_qt_console(self, evt=None):
+        """ Start a new qtconsole connected to our kernel. """
+        return connect_qtconsole(
+            self.ipkernel.connection_file, profile=self.ipkernel.profile
+        )
```
Comment (pberkes) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: Ah! Here's the bug: we never save the console in the list.

File: envisage/plugins/ipython_kernel/ipython_kernel_plugin.py (hunk 7 above)
Comment (pberkes) [inline envisage/plugins/ipython_kernel/ipython_kernel_plugin.py]: Solved (we were not killing the consoles)

File: examples/plugins/tasks/ipython_kernel/example.py (hunk 8 above)
Comment (pberkes) [inline examples/plugins/tasks/ipython_kernel/example.py]: done

File: envisage/plugins/ipython_kernel/ipython_kernel_plugin.py (hunk 3 above)
Comment (pberkes) [inline envisage/plugins/ipython_kernel/ipython_kernel_plugin.py]: done

**Comment by pberkes** [comment]
Hi Didrik, I got a bit carried away. Anyway, I added some tests and an extension point for the initial content of the namespace, so plugins can add new variables.
See if you like the changes.

**Comment by dpinte** [comment]
@pberkes thanks, will review that today.

**Comment by dpinte** [comment]
@pberkes I've remove the confirmation dialog when closing the qtconsole. Things seem to work fine. Thanks for all the cleanup.

**Comment by pberkes** [comment]
LGTM
@dpinte is there anything else you plan to do?

**Comment by dpinte** [comment]
not really except merging

**Comment by prabhuramachandran** [comment]
Nice! :)

**Pull request merged**